### ISSUE TITLE
auth: REFUSED after remote backend restarts

### ISSUE THREAD

**Issue by yosida95** [PowerDNS/pdns#8615]
- Program: Authoritative
 - Issue type: Bug report

### Short description

On PowerDNS Authoritative server deployment with remote UNIX socket backend enabled, PowerDNS sends REFUSED response to first query for each distributor thread after the backend restarts which brings UNIX socket to close.

### Environment

 - Operating system: Linux
 - Software version: 4.1.13
 - Software source: https://hub.docker.com/r/tcely/powerdns-server

### Steps to reproduce

1. launch PowerDNS Authoritative server `--launch=remote --remote-connection-string=unix:path=/path/to/backend.sock` and corresponding backend.
2. Send a query of existing resource record.
3. Restart the remote backend.
4. Repeat the query sent in 2. step.

### Expected behaviour

On both 2. and 4. step, I expect to get `NOERROR` response.

### Actual behaviour

On the 2. step I could get `NOERROR` response expectedly, but on the 4. step I got `REFUSED` unexpectedly.

**Comment by Habbie** [comment]
Do you have an example remote backend we could use to reproduce this issue?

**Comment by yosida95** [comment]
Yes, you can use this gist (written in Go) to repro the issue: https://gist.github.com/yosida95/bba384f6dd141caf6b0f1ca9b1c35a1e.

This backend responds only example.com/SOA.

```
~ ✘╹◡╹✘ dig +norec @127.0.0.1 -p 5353 example.com SOA

; <<>> DiG 9.10.3-P4-Ubuntu <<>> +norec @127.0.0.1 -p 5353 example.com SOA
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 49194
;; flags: qr aa; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1680
;; QUESTION SECTION:
;example.com.                   IN      SOA

;; ANSWER SECTION:
example.com.            3600    IN      SOA     ns.example.com. hostmaster.example.com. 1 1800 900 604800 3600

;; Query time: 0 msec
;; SERVER: 127.0.0.1#5353(127.0.0.1)
;; WHEN: Tue Dec 10 20:09:31 JST 2019
;; MSG SIZE  rcvd: 90

~ ✘╹◡╹✘ dig +norec @127.0.0.1 -p 5353 example.com SOA

; <<>> DiG 9.10.3-P4-Ubuntu <<>> +norec @127.0.0.1 -p 5353 example.com SOA
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: REFUSED, id: 11749
;; flags: qr; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1680
;; QUESTION SECTION:
;example.com.                   IN      SOA

;; Query time: 0 msec
;; SERVER: 127.0.0.1#5353(127.0.0.1)
;; WHEN: Tue Dec 10 20:09:42 JST 2019
;; MSG SIZE  rcvd: 40
```

**Comment by matthi4s** [comment]
I'm also having issues with this, seeing this problem occurring on some (not all) backend restarts, I couldn't reproduce this on the first backend connect after starting PowerDNS. 

In those failed cases, the first SOA request never reaches the backend, maybe because PowerDNS sends the first request too early or something like that. The backend then receives SOA requests for the next levels (e.g. ".com" and "."), which are answered with `false` leading to the mentioned `REFUSED`. 

Here is the output of the request/responses in the backend when this happens:
```
Client connected
REQUEST: {"method": "initialize", "parameters": {"path": "/path/to/backend.sock"}}
RESPONSE: {"result":true}

REQUEST: {"method": "lookup", "parameters": {"local": "0.0.0.0", "qname": "com.", "qtype": "SOA", "real-remote": "0.0.0.0/0", "remote": "0.0.0.0", "zone-id": -1}}
RESPONSE: {"result":false}

REQUEST: {"method": "lookup", "parameters": {"local": "0.0.0.0", "qname": ".", "qtype": "SOA", "real-remote": "0.0.0.0/0", "remote": "0.0.0.0", "zone-id": -1}}
RESPONSE: {"result":false}
```

I hope that helps to identify and solve this issue.

**Comment by yosida95** [comment]
As far as I researched, this is due to the connection abort.

When the first query reached the PowerDNS, it establishes its connection to the configured remote backend. If the remote backend restarted after that, the connection between PowerDNS and the remote backend to be broken.

Then the second query reached the PowerDNS, it tries to forward the query to the remote backend but it fails. In this case, I think the PowerDNS should raise some exceptions to answer SERVFAIL or retry the query (if it is idempotent), but the current implementation treats this case as same as an empty response (JSON: `{}`) returned.

https://github.com/PowerDNS/pdns/blob/e719aae4e8ca5b757b5af2bd1e434f2ea5ab6659/modules/remotebackend/remotebackend.cc#L194-L200
https://github.com/PowerDNS/pdns/blob/e719aae4e8ca5b757b5af2bd1e434f2ea5ab6659/modules/remotebackend/remotebackend.cc#L34-L36
https://github.com/PowerDNS/pdns/blob/e719aae4e8ca5b757b5af2bd1e434f2ea5ab6659/modules/remotebackend/unixconnector.cc#L59-L65

**Comment by rgacogne** [comment]
Wouldn't https://github.com/PowerDNS/pdns/pull/8943 fix that?

**Comment by yosida95** [comment]
@rgacogne I haven't build and test the commit, but as far as I reviewed the patch it would fixed this issue.

**Comment by Habbie** [comment]
Should be fixed in #8943.
